src/txdb.cpp

```diff
@@ -567,10 +567,13 @@ bool CBlockTreeDB::LoadBlockIndexGuts(
 
                 // ZIP 221 consistency checks
                 if (chainParams.GetConsensus().NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
-                    if (pindexNew->hashLightClientRoot != pindexNew->hashChainHistoryRoot) {
+                    // Don't enforce consistency check for headers that aren't necessarily
+                    // Heartwood-aware (such as headers we receive from non-upgraded peers
+                    // immediately after Heartwood activation).
+                    if (pindexNew->IsValid(BLOCK_VALID_CONSENSUS) && pindexNew->hashLightClientRoot != pindexNew->hashChainHistoryRoot) {
```

Comment:
This issue would also occur, if I understand correctly, for any other upgrade that changes header requirements (apart from strict loosening). It doesn't change the logic, but in order to make this clearer:

```suggestion
                // Don't enforce additional Heartwood rules for headers that aren't necessarily
                // Heartwood-aware (such as headers we receive from non-upgraded peers
                // immediately after Heartwood activation).
                if (chainParams.GetConsensus().NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)
                        && pindexNew->IsValid(BLOCK_VALID_CONSENSUS)) {
                    if (pindexNew->hashLightClientRoot != pindexNew->hashChainHistoryRoot) {
```

Reply:
This is in fact a different check; it will enforce a pre-Heartwood consistency check on post-Heartwood but invalid blocks, whereas the current PR does not attempt to enforce any consistency check on post-Heartwood invalid blocks (because they could be invalid for reasons other than being from another chain).

Reply:
Oh, I see. cf480fe402844f4c95fb57ccf3fc0568bc0aec7b added both branches of this `if`, not just the Heartwood branch. But then doesn't the same argument apply to the other branch of the `if`? There could have been invalid blocks prior to Heartwood activation that had `hashFinalSaplingRoot` set incorrectly, and that's not an error as long as they are marked as invalid.

(Why are we even keeping invalid blocks in the index, though? The block database design we inherited has never made any sense to me.)